include/ocpp/v201/component_state_manager.hpp

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2020 - 2023 Pionix GmbH and Contributors to EVerest
+
+#pragma once
+
+#include <ocpp/v201/enums.hpp>
+#include "database_handler.hpp"
+
+namespace ocpp::v201 {
+
+struct FullConnectorStatus {
+    OperationalStatusEnum individual_operational_status;
+    bool faulted;
+    bool reserved;
+    bool occupied;
+
+    ConnectorStatusEnum to_effective_status();
+};
+
+class ComponentStateManager {
+private:
+    std::shared_ptr<DatabaseHandler> database;
+
+    OperationalStatusEnum cs_individual_status;
+    std::vector<std::pair<OperationalStatusEnum, std::vector<FullConnectorStatus>>>
+        evse_and_connector_individual_statuses;
+
+    OperationalStatusEnum last_cs_effective_status;
+    std::vector<std::pair<OperationalStatusEnum, std::vector<ConnectorStatusEnum>>>
+        last_evse_and_connector_effective_statuses;
+
+    void check_evse_id(int32_t evse_id);
+    void check_evse_and_connector_id(int32_t evse_id, int32_t connector_id);
+
+public:
+    explicit ComponentStateManager(const std::map<int32_t, int32_t>& evse_connector_structure,
+                                   std::shared_ptr<DatabaseHandler> db_handler);
+
+    OperationalStatusEnum get_cs_individual_operational_status();
+    OperationalStatusEnum get_evse_individual_operational_status(int32_t evse_id);
+    OperationalStatusEnum get_connector_individual_operational_status(int32_t evse_id, int32_t connector_id);
+
+    OperationalStatusEnum get_cs_persisted_operational_status();
+    OperationalStatusEnum get_evse_persisted_operational_status(int32_t evse_id);
+    OperationalStatusEnum get_connector_persisted_operational_status(int32_t evse_id, int32_t connector_id);
+
+    bool cs_effective_status_changed();
+    bool evse_effective_status_changed(int32_t evse_id);
+    bool connector_effective_status_changed(int32_t evse_id, int32_t connector_id);
+
+    void clear_cs_effective_status_changed();
+    void clear_evse_effective_status_changed(int32_t evse_id);
+    void clear_connector_effective_status_changed(int32_t evse_id, int32_t connector_id);
```

Comment:
why are those necessary?

It is confusing to me since a state change is an event of no duration... so I would assume when a state changes (by calling one of the set_status operations) a notification is triggered (in this implementation in form of the callbacks) and that's it?  I would not quite understand the meaning of those booleans (I assumed changed ever since the last clearance; but it is weird to me that this kind of administrative burden is put on the user of this class)

Reply:
Refactored - the component state manager is now responsible for calling the callbacks as well.